File: api/src/main/java/org/delivery/api/domain/userorder/business/UserOrderBusiness.java
package org.delivery.api.domain.userorder.business;

import lombok.RequiredArgsConstructor;
import org.delivery.api.domain.store.converter.StoreConverter;
import org.delivery.api.domain.store.service.StoreService;
import org.delivery.api.domain.storemenu.converter.StoreMenuConverter;
import org.delivery.api.domain.storemenu.service.StoreMenuService;
import org.delivery.api.domain.user.model.User;
import org.delivery.api.domain.userorder.controller.model.UserOrderDetailResponse;
import org.delivery.api.domain.userorder.controller.model.UserOrderRequest;
import org.delivery.api.domain.userorder.controller.model.UserOrderResponse;
import org.delivery.api.domain.userorder.converter.UserOrderConverter;
import org.delivery.api.domain.userorder.producer.UserOrderProducer;
import org.delivery.api.domain.userorder.service.UserOrderService;
import org.delivery.api.domain.userordermenu.converter.UserOrderMenuConverter;
import org.delivery.api.domain.userordermenu.service.UserOrderMenuService;
import org.delivery.common.annotation.Business;
import org.delivery.db.userordermenu.enums.UserOrderMenuStatus;

import java.util.List;
import java.util.stream.Collectors;

@RequiredArgsConstructor
@Business
public class UserOrderBusiness {

    private final UserOrderService userOrderService;

    private final UserOrderConverter userOrderConverter;
    private final StoreMenuService storeMenuService;

    private final StoreService storeService;

    private final StoreMenuConverter storeMenuConverter;

    private final StoreConverter storeConverter;

    private final UserOrderMenuConverter userOrderMenuConverter;

    private final UserOrderMenuService userOrderMenuService;

    private final UserOrderProducer userOrderProducer;

    // 1. User, menu id
    // 2. generate userOrder
    // 3. generate userOrderMenu
    // 4. generate response
    public UserOrderResponse userOrder(User user, UserOrderRequest body) {

        var storeEntity = storeService.getStoreWithThrow(body.getStoreId());

        var storeMenuEntityList = body.getStoreMenuIdList()
                .stream()
                .map(it -> storeMenuService.getStoreMenuWithThrow(it))
                .collect(Collectors.toList());

        var userOrderEntity = userOrderConverter.toEntity(user, storeEntity, storeMenuEntityList);

        //order
        var newUserOrderEntity = userOrderService.order(userOrderEntity);

        // mapping
        var userOrderMenuEntityList = storeMenuEntityList.stream()
                .map(it -> {
                    var userOrderMenuEntity = userOrderMenuConverter.toEntity(newUserOrderEntity, it);

                    return userOrderMenuEntity;
                })
                .collect(Collectors.toList());

        // write order history
        userOrderMenuEntityList.forEach(it -> {
            userOrderMenuService.order(it);
        });

        // asynchronously send message to message queue
        userOrderProducer.sendOrder(newUserOrderEntity);

        //response
        return userOrderConverter.toResponse(newUserOrderEntity);
    }

    public List<UserOrderDetailResponse> current(User user) {

        var userOrderEntityList = userOrderService.current(user.getId());

        // handle order by order
        var userOrderDetailResponseList = userOrderEntityList.stream().map(userOrderEntity -> {

            // user ordered this menu
            var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList().stream()
                    .filter(it -> it.getStatus().equals(UserOrderMenuStatus.REGISTERED))
                    .collect(Collectors.toList());

            var storeMenuEntityList = userOrderMenuEntityList.stream()
                    .map(userOrderMenuEntity -> {
                        return userOrderMenuEntity.getStoreMenu();
                    }).collect(Collectors.toList());

            // user ordered from this store
            var storeEntity = userOrderEntity.getStore();

            return UserOrderDetailResponse.builder()
                    .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                    .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                    .storeResponse(storeConverter.toResponse(storeEntity))
                    .build()
                    ;
        }).collect(Collectors.toList());


        return userOrderDetailResponseList;
    }

    public List<UserOrderDetailResponse> history(User user) {

        var userOrderEntityList = userOrderService.history(user.getId());

        // handle order by order
        var userOrderDetailResponseList = userOrderEntityList.stream().map(userOrderEntity -> {

            // user ordered this menu
            var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList().stream()
                    .filter(it -> it.getStatus().equals(UserOrderMenuStatus.REGISTERED))
                    .collect(Collectors.toList());

            var storeMenuEntityList = userOrderMenuEntityList.stream()
                    .map(userOrderMenuEntity -> userOrderMenuEntity.getStoreMenu())
                    .collect(Collectors.toList());

            // user ordered from this store
            var storeEntity = userOrderEntity.getStore();

            return UserOrderDetailResponse.builder()
                    .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                    .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                    .storeResponse(storeConverter.toResponse(storeEntity))
                    .build()
                    ;
        }).collect(Collectors.toList());


        return userOrderDetailResponseList;
    }

    public UserOrderDetailResponse read(User user, Long orderId) {

        var userOrderEntity = userOrderService.getUserOrderWithoutStatusWithThrow(orderId, user.getId());

        var userOrderMenuEntityList = userOrderEntity.getUserOrderMenuList().stream()
                .filter(it -> it.getUserOrder().equals(UserOrderMenuStatus.REGISTERED))
                .collect(Collectors.toList());

        var storeMenuEntityList = userOrderMenuEntityList.stream()
                .map(userOrderMenuEntity -> userOrderMenuEntity.getStoreMenu())
                .collect(Collectors.toList());

        // user ordered from this store
        var storeEntity = userOrderEntity.getStore();

        return UserOrderDetailResponse.builder()
                .userOrderResponse(userOrderConverter.toResponse(userOrderEntity))
                .storeMenuResponseList(storeMenuConverter.toResponse(storeMenuEntityList))
                .storeResponse(storeConverter.toResponse(storeEntity))
                .build()
                ;
    }
}

File: api/src/main/kotlin/org/delivery/api/domain/userorder/business/UserOrderBusiness.kt
package org.delivery.api.domain.userorder.business

import org.delivery.api.common.Log
import org.delivery.api.domain.store.converter.StoreConverter
import org.delivery.api.domain.store.service.StoreService
import org.delivery.api.domain.storemenu.converter.StoreMenuConverter
import org.delivery.api.domain.storemenu.service.StoreMenuService
import org.delivery.api.domain.user.model.User
import org.delivery.api.domain.userorder.controller.model.UserOrderDetailResponse
import org.delivery.api.domain.userorder.controller.model.UserOrderRequest
import org.delivery.api.domain.userorder.controller.model.UserOrderResponse
import org.delivery.api.domain.userorder.converter.UserOrderConverter
import org.delivery.api.domain.userorder.producer.UserOrderProducer
import org.delivery.api.domain.userorder.service.UserOrderService
import org.delivery.api.domain.userordermenu.converter.UserOrderMenuConverter
import org.delivery.api.domain.userordermenu.service.UserOrderMenuService
import org.delivery.common.annotation.Business
import org.delivery.db.userordermenu.enums.UserOrderMenuStatus
import kotlin.streams.toList

@Business
class UserOrderBusiness(
    private val userOrderService: UserOrderService,
    private val userOrderConverter: UserOrderConverter,

    private val storeMenuService: StoreMenuService,
    private val storeMenuConverter: StoreMenuConverter,

    private val userOrderMenuService: UserOrderMenuService,
    private val userOrderMenuConverter: UserOrderMenuConverter,

    private val storeService: StoreService,
    private val storeConverter: StoreConverter,

    private val userOrderProducer: UserOrderProducer
) {

    companion object: Log

    fun userOrder(
        user: User?,
        body: UserOrderRequest?
    ): UserOrderResponse {
        // find store
        val storeEntity = storeService.getStoreWithThrow(body?.storeId)

        // find order menu
        val storeMenuEntityList = body?.storeMenuIdList
            ?.mapNotNull { storeMenuService.getStoreMenuWithThrow(it) }
            ?.toList()

        // order
        val userOrderEntity = userOrderConverter.toEntity(
            user = user,
            storeEntity = storeEntity,
            storeMenuEntityList = storeMenuEntityList
        ).run {
            userOrderService.order(this)
        }

        // generate user order menu list
        val userOrderMenuEntityList = storeMenuEntityList
            ?.map { it -> userOrderMenuConverter.toEntity(userOrderEntity, it) }
            ?.toList()

        userOrderMenuEntityList?.forEach { userOrderMenuService.order(it) }

        // notify order asynchronously
        userOrderProducer.sendOrder(userOrderEntity)

        return userOrderConverter.toResponse(userOrderEntity)
    }

    fun current(
        user: User?
    ): List<UserOrderDetailResponse> {
        return userOrderService.current(user?.id)
            .map { userOrderEntity ->

                val storeMenuEntityList = userOrderEntity.userOrderMenuList
                    ?.filter {userOrderMenuEntity ->
                        userOrderMenuEntity.status == UserOrderMenuStatus.REGISTERED
                    }?.map { userOrderMenuEntity ->
                        userOrderMenuEntity.storeMenu
                    }?.toList()

                UserOrderDetailResponse(
                    storeResponse = storeConverter.toResponse(userOrderEntity.store),
                    storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList),
                    userOrderResponse = userOrderConverter.toResponse(userOrderEntity)
                )
            }.toList()
    }

    fun history(
        user: User?
    ): List<UserOrderDetailResponse> {
        return userOrderService.history(user?.id)
            .map { userOrderEntity ->

                val storeMenuEntityList = userOrderEntity.userOrderMenuList
                    ?.filter {userOrderMenuEntity ->
                        userOrderMenuEntity.status == UserOrderMenuStatus.REGISTERED
                    }?.map { userOrderMenuEntity ->
                        userOrderMenuEntity.storeMenu
                    }?.toList()

                UserOrderDetailResponse(
                    storeResponse = storeConverter.toResponse(userOrderEntity.store),
                    storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList),
                    userOrderResponse = userOrderConverter.toResponse(userOrderEntity)
                )
            }.toList()
    }

    fun read(
        user: User?,
        orderId: Long?
    ): UserOrderDetailResponse {

        return userOrderService.getUserOrderWithoutStatusWithThrow(orderId, user?.id)
            .let { userOrderEntity ->

                val storeMenuEntityList = userOrderEntity.userOrderMenuList
                    ?.stream()
                    ?.filter { it.status == UserOrderMenuStatus.REGISTERED }
                    ?.map { it.storeMenu }
                    ?.toList()
                    ?: listOf()

                UserOrderDetailResponse(
                    storeResponse = storeConverter.toResponse(userOrderEntity.store),
                    storeMenuResponseList = storeMenuConverter.toResponse(storeMenuEntityList),
                    userOrderResponse = userOrderConverter.toResponse(userOrderEntity)
                )
            }

    }
}